test-utils/cli/src/lib.rs

```diff
@@ -35,6 +35,38 @@ use std::{
 };
 use tokio::io::{AsyncBufReadExt, AsyncRead};
 
+/// Similar to [`crate::start_node`] spawns a node, but works in environments where the substrate
+/// binary is not accessible with `cargo_bin("substrate-node")`, and allows customising the args
+/// passed in.
+///
+/// Helpful if you need a Substrate dev node running in the background of a project external to
+/// `substrate`.
+///
+/// The downside compared to using [`crate::start_node`] is that this method is blocking rather than
+/// returning a [`Child`]. Therefore, you may want to call this method inside a new thread.
+///
+/// # Example
+/// ```ignore
+/// // Spawn a dev node.
+/// let _ = std::thread::spawn(move || {
+///     match common::start_node_inline(vec!["--dev", "--rpc-port=12345"]) {
+///         Ok(_) => {}
+///         Err(e) => {
+///             panic!("Node exited with error: {}", e);
+///         }
+///     }
+/// });
+/// ```
+pub fn start_node_inline(args: Vec<&str>) -> Result<(), sc_service::error::Error> {
+	use sc_cli::SubstrateCli;
+
+	// Prepend the args with some dummy value because the first arg is skipped.
+	let cli_call = std::iter::once("node-template").chain(args);
+	let cli = node_template::cli::Cli::from_iter(cli_call);
```

Comment:
I did not originally get what this MR does, but now i am thinking that you can spawn the binary like we do in all the [benchmark tests](https://github.com/paritytech/substrate/blob/d006dbd1e41d3e6fd8af5a1626769d0f630e3b93/bin/node/cli/tests/benchmark_pallet_works.rs#L37).  
It will require building the node, but you are also building the code when using `start_node_inline` (and the CI can cache it anyway).

Reply:
This would still require a seperate external binary present on the path to run tests, which I'd like to avoid